### ISSUE TITLE
Freeze when reloading drouting with TLS and probing

### ISSUE THREAD

**Issue by adrien-martin** [OpenSIPS/opensips#2438]
**OpenSIPS version you are running**

```
version: opensips 2.4.9 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 815f19ad6
main.c compiled on  with gcc 6.3.0
```

**Describe the bug**

Opensips seems to slow down during `dr reload` with 30+ TLS gateways.

dr reload always finishes properly but can last 10s and more.
During this time we noticed that outgoing/incoming probings and registrations can fail.

TCP gateways seems not to have this problem.

**Additional context**

TCP and TLS timeout are set to 1s (edit, no impact).
We have about 600 gateways 30 TLS gateways and 400 rules.

**To Reproduce**

30+ TLS gateways.
Some traffic.
Multiple `opensipsctl dr reload` until it occurs.

**Expected behavior**

No slow down.
Short drouting reloading time.

**Relevant System Logs**

WARNING:core:utimer_ticker: utimer task <tm-utimer> already scheduled for 2335190 ms (now 2335200 ms), it may overlap..

**OS/environment information**
Debian 9
Debian packages compiled from 2.4.9 tag with some backported commits:
* 2854afbf8,
* 58f42976e,
* 07bdb586c.

An additional custom patch that disable drouting gateways when the socket is not local.
At first we thought is was a problem with non local sockets, this patch help in this regard but only because it reduce the number of gateways.

drouting.c for the backtrace: [drouting.c.tar.gz](https://github.com/OpenSIPS/opensips/files/6163184/drouting.c.tar.gz).

**Backtrace**

[BTDroutingFreezeTLS.txt](https://github.com/OpenSIPS/opensips/files/6163172/BTDroutingFreezeTLS.txt)

**Comment by Nestorfish** [comment]
[opensips_ps.txt](https://github.com/OpenSIPS/opensips/files/6165889/opensips_ps.txt)
opensipsctl ps for the already given opensipsctl trap

**Comment by Nestorfish** [comment]
Attached is a forced debug of locks occuring inside drouting.c, occuring when a dr reload is running.
Logs are set to error, to avoid setting OpenSIPS to debug and overloading its logs.
We can reproduce this behavior out of production, so do not hesitate to ask for real debug logs if you want.
[Uploading dr_reload_locks_debug_anon.txt…]()

**Comment by adrien-martin** [comment]
Opensips 3.1.1 (no patches) seems to behave in the same way, at least for the `dr_reload` duration increase.

**Comment by bogdan-iancu** [comment]
@adrien-martin , what probing mode do you have set for the GWs? Also, are the GWs specified as IPs or FQDNs ?

**Comment by adrien-martin** [comment]
Hello, `probe_mode` is set to 2 (permanent probing).
All the TLS gateways have a FQDN name, the SQL `address` field is like `<fqdn>;<port>;transport=tls.
For UDP gateways plain IP are used with the same format.

**Comment by bogdan-iancu** [comment]
I'm asking as the actual `dr_reload` task happens outside the SIP workers, so it should not interfere with the performance on SIP handling. The only thing to slow the `dr_reload` is the DNS lookup on the GW's name - could you use IP's instead of FQDN's and try again the reload ?

**Comment by adrien-martin** [comment]
Ah, I didn't thought about DNS performance.
Just ran some tests with IP instead of domain names, it does not change the behavior.

We also check the DNS traffic with domain names, the flow of queries/responses is clean at this moment.
The responses are fast, we use a DNS cache on a nearby host.
The latency for TTL refreshs is quite low too.

**Comment by bogdan-iancu** [comment]
@adrien-martin , without the FQDN's, there should be nothing to slow the `dr_reload`. And the fact that the GWs are TLS or not, again, should have no impact, as there is no singling or protocol related ops during the reload.
BTW, you mentioned 30+ gws, but what about the number of rules? how many do you have ?

Also, what you can do it do try to do multiple traps during reload (the first one you uploaded is not relevant).  As the trap cmd may take long (as it takes the bts from all opensips procs), you can try a manual trap, only on the reloading process. If you use the FIFO as MI backed, do an MI `ps`, locate the pid of the FIFO process and manually run gdb (use `-ex bt` to avoid the gdb console/interactive mode) several times, until the reload ends.

**Comment by adrien-martin** [comment]
We have about 20 rules using 20 carriers and 400 rules using gateways without carriers.
  All the 30+ TLS gateways are not par of a carrier.
  
 Here are 3 backtraces, during a 6s `dr_reload`.
The first two are similar to backtrace in the first message, the third is similar to the backtraces when `dr_reload` was finished.
[bts.tar.gz](https://github.com/OpenSIPS/opensips/files/6249585/bts.tar.gz)

**Comment by bogdan-iancu** [comment]
@adrien-martin , thanks for the info. Both the first 2 BTs, point to the same place, to a `usleep()` inside `dr_reload_data_head()`. The only occurrence on that `usleep` is via `lock_start_write()` - a "multiple-reads one-writer" safe guard.
Basically the "reload" takes so long not because of the actual loading, but because it is waiting to safely replace the old data with the new data. So the reload is waiting for (a) another reload (but this should go fast) or (b) some readers.
And the questions are:
1) do you do multiple reloads?
2) while reloading, do you actually have traffic (calling do_routing()) ?

**Comment by adrien-martin** [comment]
Multiple reloads can occur at the same time in production, not in these tests though.
We do not really need concurrent reloads, but we are not trying to prevent it at the moment (`no_concurrent_reload` not set and no specific control on the programs using `dr_reload`).

No drouting traffic during the tests but a lot of probing.

**Comment by Nestorfish** [comment]
@bogdan-iancu 
It is dr_prob_handler() read lock on it->ref_lock that lasts for multiple seconds each time it runs. It is the same lock that dr_reload_data_head tries to acquire for writing.

**Comment by bogdan-iancu** [comment]
@Nestorfish , that;s a good point, the probing handler is also acting as a reader, so it may block the reloader writing until the the probes are sent. Switch to no-probing mode and see if the problem goes away....

**Comment by Nestorfish** [comment]
Probing is a condition to reproduce the problem, effectively.

**Comment by Nestorfish** [comment]
The probing is slowed down by TLS gateways that are slow, nothing OpenSIPS can help here, unless doing some dirty tricks (working on a copy of GWs data, then locking to update existing GWs ?).

However, I do not understand why we have such impact on regular trafic when dr reload occurs (not every time, but often) :
 - not responding to UDP requests (monitoring process often restarted our OpenSIPS instances while dr reload was in progress, and customers also noticed it as their equipment logged our proxies not responding, timestamps corresponding with reloads). I observed this with another utility sending OPTIONS which did not get response. We are currently not able to reproduce this on our non-production machine.
 - utimer task <tm-utimer> that hangs until reload finishes (another lock issue ?)

**Comment by Nestorfish** [comment]
Update: utimer task tm-utimer hangs as soon as reloading occurs during probing, it is independent of absence of processing of UDP traffic.

**Comment by bogdan-iancu** [comment]
Here is the explanation - the DR pinging routine acts as a very very slow reads - it keep the DR data marked for reading while trying to send the pings. If there is a reload in the mean while, this will block and wait until all the ongoing read ops will terminate. So the reload will wait for the pinging routine to stop. While there is waiting writer (the reload in our case), no other NEW read operation is allowed (even if other read ops are already in progress), but they will be postpone for after the write op will complete.  This is done to be sure that the writer has priority over the multiple readers.
But this creates a chain of waiting reader (sip traffic) is waiting for writer (reloader) which waits for another reader (the pinger).
Does it make sense ?

**Comment by adrien-martin** [comment]
I think this makes sense.
I guess it's the same lock for UDP, TCP and TLS probing, maybe reducing the number of probed UDP gateways could help a bit mitigating the problem.

**Comment by Nestorfish** [comment]
OK, I missed the point that other workers will soon or later call do_routing then also block, until probing is done.

**Comment by bogdan-iancu** [comment]
It is a single lock, protecting the dr data, it has nothing to do with the transport protocol.

I think the best "fix" here will be to avoid holding the lock in the probing routine while actually sending the probes out. In this ways, whatever delay may appear while doing the signaling will not create the starvation for the access to the DR data.

**Comment by adrien-martin** [comment]
@github-actions please unstale.

**Comment by bogdan-iancu** [comment]
@adrien-martin , I have this on my TODO list, unfortunately not with a high-priority. Still, it is there :)

**Comment by adrien-martin** [comment]
Sorry for the noise, i was afraid the stale bot could make some issues get lost.

**Comment by adrien-martin** [comment]
@github-actions please unstale.

**Comment by adrien-martin** [comment]
Please do not close.

**Comment by bogdan-iancu** [comment]
@adrien-martin , I made a fix for this on master branch (3.3), but it is not a critical issue in order to do a backport. Nevertheless, this is something you can do by yourself if this is a big need for you.

**Comment by adrien-martin** [comment]
Thanks for the commit and for your time.
No worry for the backport, i can build it indeed, and we need to use newer versions at some point anyway.
